ffi/memorymanager.h

```diff
@@ -0,0 +1,203 @@
+        /// address following \p NearBlock is not so aligned, it will be rounded
+        /// up to the next allocation granularity boundary.
+        ///
+        /// \r a non-null MemoryBlock if the function was successful, otherwise
+        /// a null MemoryBlock with \p EC describing the error.
+        virtual sys::MemoryBlock
+        allocateMappedMemory(AllocationPurpose Purpose, size_t NumBytes,
+                             const sys::MemoryBlock *const NearBlock,
+                             unsigned Flags, std::error_code &EC) = 0;
+
+        /// This method sets the protection flags for a block of memory to the
+        /// state specified by \p Flags.  The behavior is not specified if the
+        /// memory was not allocated using the allocateMappedMemory method. \p
+        /// Block describes the memory block to be protected. \p Flags specifies
+        /// the new protection state to be assigned to the block.
+        ///
+        /// If \p Flags is MF_WRITE, the actual behavior varies with the
+        /// operating system (i.e. MF_READ | MF_WRITE on Windows) and the target
+        /// architecture (i.e. MF_WRITE -> MF_READ | MF_WRITE on i386).
+        ///
+        /// \r error_success if the function was successful, or an error_code
+        /// describing the failure if an error occurred.
+        virtual std::error_code
+        protectMappedMemory(const sys::MemoryBlock &Block, unsigned Flags) = 0;
+
+        /// This method releases a block of memory that was allocated with the
+        /// allocateMappedMemory method. It should not be used to release any
+        /// memory block allocated any other way. \p Block describes the memory
+        /// to be released.
+        ///
+        /// \r error_success if the function was successful, or an error_code
+        /// describing the failure if an error occurred.
+        virtual std::error_code releaseMappedMemory(sys::MemoryBlock &M) = 0;
+
+        virtual ~MemoryMapper();
+    };
+
+    /// Creates a LlvmliteMemoryManager instance with \p MM as the associated
+    /// memory mapper.  If \p MM is nullptr then a default memory mapper is used
+    /// that directly calls into the operating system.
+    LlvmliteMemoryManager(MemoryMapper *MM = nullptr);
+    LlvmliteMemoryManager(const LlvmliteMemoryManager &) = delete;
+    void operator=(const LlvmliteMemoryManager &) = delete;
+    ~LlvmliteMemoryManager() override;
+
+    /// Allocates a memory block of (at least) the given size suitable for
+    /// executable code.
+    ///
+    /// The value of \p Alignment must be a power of two.  If \p Alignment is
+    /// zero a default alignment of 16 will be used.
+    uint8_t *allocateCodeSection(uintptr_t Size, unsigned Alignment,
+                                 unsigned SectionID,
+                                 StringRef SectionName) override;
+
+    /// Allocates a memory block of (at least) the given size suitable for
+    /// executable code.
```

Comment:
Nit: executable code -> data.
(looks a copy-paste typo from the previous method)